### ISSUE TITLE
time_bound variable requirement when computing monthly anomaly, annual mean, etc

### ISSUE THREAD

**Issue by andersy005** [NCAR/esmlab#7]
@matt-long, while looking at 
<details>
<summary>def compute_mon_anomaly(dsm) function</summary>

    tb_name,tb_dim = time_bound_var(dsm)
    grid_vars = [v for v in dsm.variables if 'time' not in dsm[v].dims]
    variables = [v for v in dsm.variables if 'time' in dsm[v].dims and v not in ['time',tb_name]]

    # save attrs
    attrs = {v:dsm[v].attrs for v in dsm.variables}
    coords = {v:dsm[v].attrs for v in dsm.variables}
    encoding = {v:{key:val for key,val in dsm[v].encoding.items()
                   if key in ['dtype','_FillValue','missing_value']}
                for v in dsm.variables}

    #-- compute time variable
    time_values_orig = dsm.time.values
    date = cftime.num2date(dsm[tb_name].mean(tb_dim),
                           units = dsm.time.attrs['units'],
                           calendar = dsm.time.attrs['calendar'])
    dsm.time.values = date
    if len(date)%12 != 0:
        raise ValueError('Time axis not evenly divisible by 12!')

    #-- compute anomaly
    ds = dsm.drop(grid_vars).groupby('time.month') - 
    dsm.drop(grid_vars).groupby('time.month').mean('time')
    ds.reset_coords('month',inplace=True)

    #-- put grid_vars back
    ds = xr.merge((ds,dsm.drop([v for v in dsm.variables if v not in grid_vars])))
    ds.time.values = time_values_orig

    attrs['month'] = {'long_name':'Month'}

    # put the attributes back
    for v in ds.variables:
        ds[v].attrs = attrs[v]

    # put the encoding back
    for v in ds.variables:
        if v in encoding:
            ds[v].encoding = encoding[v]

    return ds

</details>

<br>

I noticed that you are finding the `time_bound` variable first. I am wondering whether this is a prerequisite condition for computing monthly anomaly? In other words, can we generalize this function? Is `time_bound` variable earth-system models specific? If not, can we just make it an optional argument? The goal would be to have a somewhat general function that one can use with most datasets.

**Comment by matt-long** [comment]
I think we pull out the parts of this function that handle time and consider these operations to have already been applied prior to invoking this function.

Just as background, there are two issues with time: 
1. Interpreting the metadata converting to a date object has historically been problematic thru `xarray`; many issues have been resolved and `cftime` provides functional time processing—though I still have trouble plotting timeseries with `matplotlib` and `cftime` objects as an axis.
1. CESM produces files with the time stamp at the end of an averaging period; this causes the monthly mean and other operations to fail because end-of-month or end-of-year values are pushed to the next (wrong) interval. A general solution to this is to average the `time_bound` field, if provided.

So in summary again, I think we should isolate the special time treatment tools elsewhere, which will clean things up here.

**Comment by andersy005** [comment]
@matt-long, Does the following workflow make sense?

- Call necessary operation function (`compute_climatology()`, etc). These operation functions would need to have some **flags** as optional arguments  that will allow us to do the necessary ops such as getting time bound var for CESM datasets for instance prior doing anything else. 
- Retrieve original attributes and metadata 
- Handle time related issues in one or two separate functions, `get_time_bound_var()`, `compute_time_var()`
-  Apply the necessary operation (climatology, anomaly, etc). 
- Update the resulting dataset/datarray with original attributes/metadata
- Return xarray dataset 


Here's a rough draft of pseudocode for this:

```python

def time_bound_utils(dsm):
    tb_name,tb_dim = time_bound_var(dsm)

    grid_vars = [v for v in dsm.variables if 'time' not in dsm[v].dims]
    variables = [v for v in dsm.variables if 'time' in dsm[v].dims and v not in ['time',tb_name]]

    # save attrs
    attrs = {v:dsm[v].attrs for v in dsm.variables}
    encoding = {v:{key:val for key,val in dsm[v].encoding.items()
                   if key in ['dtype','_FillValue','missing_value']}
                for v in dsm.variables}

    #-- compute time variable
    date = cftime.num2date(dsm[tb_name].mean(tb_dim),
                           units = dsm.time.attrs['units'],
                           calendar = dsm.time.attrs['calendar'])
    dsm.time.values = date
    if len(date)%12 != 0:
        raise ValueError('Time axis not evenly divisible by 12!')

def update_attrs(ds):
    attrs['time'] = {'long_name':'Month','units':'month'}
    del encoding['time']

    # put the attributes back
    for v in ds.variables:
        ds[v].attrs = attrs[v]

    # put the encoding back
    for v in ds.variables:
        if v in encoding:
            ds[v].encoding = encoding[v]


def compute_mon_climatology(dsm, get_time_bound_var=False, **kwargs):
    '''Compute a monthly climatology'''
   if get_time_bound_var:
         time_bound_utils(dsm)
    #-- compute climatology
    ds = dsm.drop(grid_vars).groupby('time.month').mean('time').rename({'month':'time'})

    #-- put grid_vars back
    ds = xr.merge((ds,dsm.drop([v for v in dsm.variables if v not in grid_vars])))
    
    # Update attributes/metadata by calling `update_attrs()`
    update_attrs(ds)  
    return ds
```

**Comment by andersy005** [comment]
For this step:

```python
#-- compute time variable
    date = cftime.num2date(dsm[tb_name].mean(tb_dim),
                           units = dsm.time.attrs['units'],
                           calendar = dsm.time.attrs['calendar'])
    dsm.time.values = date
    if len(date)%12 != 0:
        raise ValueError('Time axis not evenly divisible by 12!')
```
We would need to compute time variable when we have the `time_bound_var`, otherwise we just use the provided `time` variable in the original dataset, right?
